qp_spades/qp_spades.py

```diff
@@ -6,12 +6,134 @@
 # The full license is in the file LICENSE, distributed with this software.
 # -----------------------------------------------------------------------------
 import pandas as pd
-
-from os.path import basename
+from os import environ
+from os.path import join, exists
+from glob import glob
+from subprocess import run, PIPE
 
 from qiita_client import ArtifactInfo
 from qiita_client.util import system_call
 
+# resources per job
+WALLTIME = '200:00:00'
+MAX_RUNNING = 8
+FINISH_WALLTIME = '10:00:00'
+FINISH_MEMORY = '4g'
+
+
+def spades_to_array(directory, output_dir, prefix_to_name, url,
+                    job_id, params):
+    environment = environ["ENVIRONMENT"]
+    ppn = params["threads"]
+    memory = params["memory"]
+
+    # 1. create file list
+    files = []
+    for prefix, sample_name in prefix_to_name.items():
+        fps = sorted(glob(join(directory, prefix + '*')))
+        # this should never occur but better to confirm
+        if len(fps) != 2:
+            mgs = f'There are multiple files that match "{prefix}"'
+            return False, None, mgs
+        files.append('\t'.join([fps[0], fps[1], prefix]))
+
+    # 2. format main comand
+    # cmd = (f'{spades_cmd} > {out_dir}/%s.spades.log 2>&1 && '
+    #            f'mv {out_dir}/{sample_name}/scaffolds.fasta {out_fp}' % (
+    #             fwd_fp, rev_fp, sample_name, sample_name))
+    command = (
+        f'spades.py --{params["type"]} -t {ppn} -m {memory} '
+        f'-k {params["k-mers"]} -o $OUTDIR/$SNAME')
+    if params['merging'].startswith('flash '):
+        # get read lenght quickly; note that we are going to assume
+        # that (1) the forward and reverse are the same lenght and (2)
+        # all file pairs have the same lenght so only calculate once
+        fp = glob(join(directory, list(prefix_to_name)[0] + '*'))[0]
+        std_out, std_err, return_value = system_call(
+                f'gunzip -c {fp} | head -n 2')
+        if return_value != 0:
+            error_msg = (f"Error uncompressing: {fp}\n"
+                         f"Std out: {std_out}\nStd err: {std_err}\n")
+            return False, None, error_msg
+        read_length = len(std_out.split('\n')[1])
+        percentage = int(params['merging'][6:-1])/100
+        overlap = int(read_length * percentage)
+
+        command = (
+            # flash
+            f'flash --threads {ppn} --max-overlap={overlap} '
+            '--output-directory $OUTDIR '
+            '--output-prefix="$SNAME" ${FWD} ${REV} '
+            '--max-mismatch-density=0.1 > $OUTDIR/${SNAME}.flash.log 2>&1'
+            ' && '
+            # spades
+            f'{command} '
+            '--merge $OUTDIR/${SNAME}.extendedFrags.fastq '
+            '-1 $OUTDIR/${SNAME}.notCombined_1.fastq '
+            '-2 $OUTDIR/${SNAME}.notCombined_2.fastq')
+    else:
+        command = '%s -1 ${FWD} -2 ${REV}' % command
+
+    # 3. create qsub for array submission
+    num_samples = len(prefix_to_name)
```

Comment:
Will there ever be more than 1024 samples?

Reply:
I really, really hope that no. I was even thinking in limiting the number of samples in a preparation via code; I'll raise this question in the qiita.admin's email.

Reply:
Okay thanks. If it's expected to be exceptional, I recommend adding an explicit check and raising an error. The job arrays cannot handle more than 1024 tasks